### ISSUE TITLE
Race condition while adding local packages to cache

### ISSUE THREAD

**Issue by julianduque** [npm/npm#12669]
We are experiencing a weird problem while installing local (`file:`) packages with npm. It happens more frequently with `npm@3` and local scoped packages, but it's totally random.

```
⬢ npm run clean && npm cache clean && npm i

> npm-bug@1.0.0 clean /Users/julianduque/Sandbox/npm-bug
> rm -rf node_modules ns-*/node_modules

npm ERR! addLocal Could not install /Users/julianduque/Sandbox/npm-bug/ns-router
npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/julianduque/.nvm/versions/node/v4.4.3/bin/node" "/Users/julianduque/.nvm/versions/node/v4.4.3/bin/npm" "i"
npm ERR! node v4.4.3
npm ERR! npm  v3.8.9
npm ERR! path /Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.783164638
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall lstat

npm ERR! enoent ENOENT: no such file or directory, lstat '/Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.783164638'
npm ERR! enoent ENOENT: no such file or directory, lstat '/Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.783164638'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/julianduque/Sandbox/npm-bug/npm-debug.log
```

Debugging this issue it happens in the following function call:

https://github.com/npm/npm/blob/master/lib/cache/add-local.js#L110

So, while `fs-write-stream-atomic` is still creating / cleaning up the temporary package file, `chownr` reads the temporary file on this function call: https://github.com/isaacs/chownr/blob/master/chownr.js#L8 but when is doing the `lstat` here: https://github.com/isaacs/chownr/blob/master/chownr.js#L18 the file doesn't exist and it crashes.

Adding some output to `fs-write-stream-atomic` and `chownr` to track down the error:
```
* npm:thenPack  /Users/julianduque/.npm/@ns/router/1.0.0/package.tgz
* npm:thenPack  /Users/julianduque/.npm/@ns/root/1.0.0/package.tgz
- fs-write-stream-atomic: /Users/julianduque/.npm/@ns/router/1.0.0/package.tgz.2454299972
- fs-write-stream-atomic:  /Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.2894581287
~ npm:moveIntoPlace  /Users/julianduque/.npm/@ns/router/1.0.0/package.tgz.2454299972 1462994637760
~~ chownr childrens:  1462994637761 /Users/julianduque/.npm/@ns
[ 'root', 'router' ]
~~ chownr childrens:  1462994637762 /Users/julianduque/.npm/@ns/root
[ '1.0.0' ]
~~ chownr childrens:  1462994637762 /Users/julianduque/.npm/@ns/router
[ '1.0.0' ]
~ fs-write-stream-atomic:moveIntoPlace  /Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.2894581287 1462994637762
~~ chownr childrens:  1462994637762 /Users/julianduque/.npm/@ns/root/1.0.0
[ 'package', 'package.tgz.2894581287' ] // ---> booom, temporary file
~~ chownr childrens:  1462994637762 /Users/julianduque/.npm/@ns/router/1.0.0
[ 'package', 'package.tgz' ] // ---> this looks good though
npm ERR! addLocal Could not install /Users/julianduque/Sandbox/npm-bug/ns-router
....
npm ERR! enoent ENOENT: no such file or directory, lstat '/Users/julianduque/.npm/@ns/root/1.0.0/package.tgz.2894581287'
```

It seems like a race condition in `fs-write-stream-atomic` or in the add-local logic, but I'm not sure how to fix it to submit a PR.

I have a small project to reproduce this issue, let me know if you need it for debugging purposes.

@othiym23 - copying you since you give us the clue about `fs-write-stream-atomic`.

#### Possibly Related

* https://github.com/npm/npm/issues/10656

**Comment by othiym23** [comment]
This is fantastic sleuthing, @julianduque! Thanks so much for all the work you've put into this!

It seems to me that the resolution is going to have to change the caching code such that `chownr` isn't called while `fs-write-stream-atomic` is in the middle of doing its business.

The problem is not in `fs-write-stream-atomic` – it's actually quite a bit more subtle than that. Much of the caching logic relies on a utility function called `getCacheStat()`. Despite its name, it works by trying to create the cache folder with the correct permissions every time it's run. The utility modules it in turn calls [`mkdirp`](http://npm.im/mkdirp) and [`chownr`](http://npm.im/chownr), which recursively create / set privileges on paths. In the case in which you're packing multiple packages into the cache simultaneously (which isn't that uncommon, especially if you're using packages installed from hosted Git repositories, or local dependencies), it's possible to hit a situation where `getCacheStat()` is calling `chownr()` when a package is just about finished being packed, so `fs-write-stream-atomic` renames the file to its proper final location before `chownr` gets to it.

The relevant code:

- https://github.com/npm/npm/blob/master/lib/cache/add-local.js#L92
- https://github.com/npm/npm/blob/master/lib/cache/get-stat.js
- https://github.com/npm/npm/blob/master/lib/utils/correct-mkdir.js#L113-L117

There are two obvious solutions, of varying degrees of hackishness:

1. Modify `getCacheStat()` to not be quite so prolific in calling `chownr()` – in a way, the current behavior makes the cache self-healing when weird stuff happens to it, but it's at best unnecessary.
2. Modify the callback to the `chownr()` in `correct-mkdir` (or `getCacheStat()`) to ignore `ENOENT` errors, which is something the code base does in a few other places.

Writing tests for this would require some fairly involved mocking and a somewhat contrived purposeful unleashing of Zalgo, but if somebody's up for a challenge, such tests would be very important in guarding against future regressions in a very hot piece of the caching code.

Thanks again, Julian! And to @brycebaril for alerting me to this issue in the first place!

**Comment by julianduque** [comment]
Thanks @othiym23 for your reply.

I tried what you suggested in 2. but, it didn't work, the race condition is still present. I changes this line https://github.com/npm/npm/blob/master/lib/cache/add-local.js#L110 to the following and it fixes the issue:

``` js
chownr(made || tgz, cs.uid, cs.gid, function (er) {
  if (er && er.code === 'ENOENT') return wrapped()
  wrapped(er)
})
```

As you say the challenge will be to write tests for this case.

**Comment by othiym23** [comment]
@julianduque Changing the `chownr()` in `lib/cache/add-local.js` won't fix the problem, because that's not the call that's crashing – it's the `chownr()` in `lib/utils/correct-mkdir.js` that's racing.

**Comment by julianduque** [comment]
@othiym23 maybe we are looking in different places but in my reproduction case the crash is on the `chownr` in `lib/cache/add-local.js`

**Comment by othiym23** [comment]
@julianduque You're right! I had to dig even further, and there's a very important line in the docs for `mkdirp` I was missing:

> `cb(err, made)` fires with the error or the first directory made that had to be created, if any.

In one of those cases (it doesn't really matter much which for the purposes of the race), it's going to be `~/.npm/@ns`, which allows it to see the in-progress tarball with temporary path. Does this bug only manifest when running with a clean npm cache? If not, there's probably more going on here than we're seeing.

The path I pointed out above is _also_ a race – it may be harder or easier to hit, but the `chownr()` in `lib/utils/correct-mkdir.js` can hit the exact same path, with the exact same problem.

 I think patching both those calls to not care about `ENOENT` is probably the safest solution to this problem. Longer term, it might make sense to reduce the number of calls to `chownr()` to one at the end and / or beginning of accessing the cache per CLI process, but that feels a little more fraught than we should take on right now.

Thanks for your patience and help, Julian!

**Comment by julianduque** [comment]
@othiym23 yes! it happens only after a `npm cache clean`, the workaround we are using is:

`$ npm install || npm install`

And it will always succeed on the 2nd try, so it's what you pointed before!

**Comment by zkat** [comment]
Hey, I'm working on something very similar over at #13023 that was affecting Windows users but that involves races in this same area of the code. I'd like to merge your patch into mine, which I'm working on writing tests for this week, and then land them together.

**Comment by julianduque** [comment]
@zkat great! I didn't added tests due to the difficulty of a reproduction case, but having this fixed would be awesome :) Thanks! <3
